File: examples/cs/gmsub_disp_rr/gmsub_disp_rr.cs
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/*! \file gmsub_disp_rr.cs
 *  A C# example subscriber using the GMSEC API dispatcher. The callback
 *  is registered with the connection. A seperate thread, transmits a request
 *  and waits for a reply. THe Dispatcher is then started, which launches its
 *  own thread.
*/


namespace gmsub_disp_rr
{
	using GMSEC.API;
	using GMSEC.API.UTIL;
	using GMSEC.API.Example.Common;
	using System;
	using System.Diagnostics;
	using System.Collections.Generic;


	class PublishCallback : Callback
	{
		public override void OnMessage(Connection conn, Message message)
		{
			Log.Info("[PublishCallback.OnMessage] Received Message:\n" + message.ToXML());
		}
	}


	class MyReqReplyCallback : ReplyCallback
	{
		public override void OnReply(Connection conn, Message request, Message reply)
		{
			Log.Info("[MyReqReplyCallback.OnReply] Received Reply Message:\n" + reply.ToXML());
		}

		public override void OnEvent(Connection conn, Status status, Connection.ConnectionEvent connEvent)
		{
			if (status.IsError())
			{
				Log.Error("[MyReqReplyCallback:OnEvent] Status=" + status.Get() + " [event=" + connEvent + "]");
			}
		}
	}


	class gmsub_disp_rr : GmsecExample
	{
		private Connection conn;
        private List<SubscriptionInfo> info = new List<SubscriptionInfo>();
 

		public gmsub_disp_rr()
		{
		}

		private void PrintUsage()
		{
			Console.Out.WriteLine(
				"\nusage: gmsub_cs_disp_rr connectiontype=<middleware> " +
				"[ <parameter>=<value> ]\n" + 
				"\n\tNote that the parameter 'connectiontype' is required. " +
				"\n\tThe rest of other parameters are optional.\n" +
				"\n\tserver=<server name> " +
				"(required if middleware is not bolt/MB locally)" +	
				"\n\tsubject=<subject name>" +
				"\n\tmsg_timeout_ms=<timeout period (milliseconds)>" +
				"\n\tprog_timeout_s=<timeout period (seconds)>" +
				"\n\tloglevel=<log level>" +
				"\n\tcfgfile=<config_filename.xml>" +
				"\n\nFor more information, see API User's Guide\n");
		}


		private void CleanUp()
		{
			if (conn != null)
			{
                for (int i = info.Count - 1; i >= 0; i--)
                {
                    Log.Info("Unsubscribing to " + info[i].GetSubject());
                    var temp = info[i];
                    conn.Unsubscribe(ref temp);
                    info.RemoveAt(i);
                }
                
                conn.Disconnect();

				Connection.Destroy(ref conn);
			}
		}


		private void Run(string[] args)
		{
			Config config = new Config(args);

			Initialize(config);

			AddToConfigFromFile(config);

			if (IsOptionInValid(config, args, false))
			{
				PrintUsage();
				return;
			}

			config.AddValue("isThreaded", "true");

			// Output GMSEC API version
			Log.Info(Connection.GetAPIVersion());

			try
			{
				// Create the connection
				conn = Connection.Create(config);

				// Connect
				conn.Connect();

				// Output the middleware info
				Log.Info("Middleware version = " + conn.GetLibraryVersion());


				// Determine the subjects to listen too
				List<string> subjects = DetermineSubjects(config);

				//Determine Timeout Periods
				int msg_timeout_ms = Get(config, "MSG_TIMEOUT_MS", (int) GMSECWaitDefs.WAIT_FOREVER);
				int prog_timeout_s = Get(config, "PROG_TIMEOUT_S", (int) GMSECWaitDefs.WAIT_FOREVER);

				// Subscribe
				PublishCallback cb = new PublishCallback();
				for (int i = 0; i < subjects.Count; i++)
				{
					Log.Info("Subscribing to "+subjects[i]);
					info.Add(conn.Subscribe(subjects[i], cb));
				}

				Log.Info("Starting AutoDispatch");
				conn.StartAutoDispatch();

				string subject = "GMSEC.TEST.PUBLISH.NOREPLY";
				if (subjects[0] != "GMSEC.>")
				{
					subject = subjects[0];
				}

				// Create request message
				Message request = new Message(subject, Message.MessageKind.REQUEST);

				// Display XML representation of the request message
				Log.Info("Sending Request Message:\n" + request.ToXML());

				// Issue request using custom ReplyCallback
				MyReqReplyCallback rcb = new MyReqReplyCallback();
				conn.Request(request, msg_timeout_ms, rcb, 0);

				if (prog_timeout_s == (int)GMSECWaitDefs.WAIT_FOREVER)
				{
					Console.Out.WriteLine("Press the Enter Key to Exit");
					Console.ReadKey();
				}
				else
				{
					System.Threading.Thread.Sleep(prog_timeout_s * 1000);

					Log.Info("Program Timeout Reached!");
				}

				Log.Info("Stopping AutoDispatch");
				conn.StopAutoDispatch();
			}
			catch (GMSEC_Exception e)
			{
				Log.Error("Exception: " + e.ToString());
			}
			catch (Exception e)
			{
				Log.Error("Exception: " + e.ToString());
			}
			finally
			{
				CleanUp();
			}
		}


		static void Main(string[] args)
		{
			new gmsub_disp_rr().Run(args);
		}
	}
}

File: examples/cpp/gmsub_disp_rr/gmsub_disp_rr.cpp
/*
 * Copyright 2007-2016 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file gmsub_disp_rr.cpp
 *
 * This file contains an example subscriber using the GMSEC standard C++ API dispatcher.
 * The callback is registered with the Connection.  A separate thread, transmits a request
 * and waits for a reply.  The Dispatcher is then started, which launches in its own
 * thread.  We're looking for contention issues with the message queue, you should be
 * able to receive messages while still waiting for a reply.
 */

#include "../example.h"

#include <iostream>
#include <string>
#include <vector>

using namespace gmsec::api;


class PublishCallback : public Callback
{
public:
	virtual void CALL_TYPE onMessage(Connection& connection, const Message& message)
	{
		GMSEC_INFO << "[PublishCallback::onMessage] Received:\n" << message.toXML();
	}
};


class RequestCallback : public ReplyCallback
{
public:
	virtual void CALL_TYPE onReply(Connection& connection, const Message& request, const Message& reply)
	{
		GMSEC_INFO << "[RequestCallback::onReply] Received Reply:\n" << reply.toXML();
	}

	virtual void CALL_TYPE onEvent(Connection& connection, const Status& status, Connection::ConnectionEvent event)
	{
		GMSEC_WARNING << "[RequestCallback::onEvent] Status=" << status.get()
		              << " event=" << event;
	}
};


class gmsub_disp_rr
{
public:
	gmsub_disp_rr(Config& config);
	~gmsub_disp_rr();
	bool run();

private:
	typedef std::vector<std::string> Subjects;

	Config&     config;
	Connection* connection;
	Subjects    subjects;
	SubscriptionInfo** info;
};


gmsub_disp_rr::gmsub_disp_rr(Config& c)
	: config(c),
	  connection(0),
	  subjects()
{
	/* Initialize config */
	example::initialize(c);
}


gmsub_disp_rr::~gmsub_disp_rr()
{
	if (connection)
	{
		for (size_t i = 0; i < subjects.size(); ++i)
		{
			GMSEC_INFO << "Unsubscribing from " << subjects[i].c_str();
			connection->unsubscribe(info[i]);
		}
		delete[] info;
		connection->disconnect();
		Connection::destroy(connection);
	}

	Connection::shutdownAllMiddlewares();
}


bool gmsub_disp_rr::run()
{
	bool success = true;

	//o output GMSEC API version
	GMSEC_INFO << Connection::getAPIVersion();

	try
	{
		config.addValue("isThreaded", "true");

		//o Create the Connection
		connection = Connection::create(config);

		//o Connect
		connection->connect();

		//o Determine the subjects to listen to
		example::determineSubjects(config, subjects);

		//o Output the middleware information
		GMSEC_INFO << "Middleware version = " << connection->getLibraryVersion();

		//o Subscribe using Callback sub-class
		PublishCallback cb;

		info = new SubscriptionInfo*[subjects.size()];

		for (size_t i = 0; i < subjects.size(); ++i)
		{
			GMSEC_INFO << "Subscribing to " << subjects[i].c_str();
			info[i] = connection->subscribe(subjects[i].c_str(), &cb);
		}

		GMSEC_INFO << "Starting AutoDispatch";
		connection->startAutoDispatch();

		RequestCallback rcb;

		int msg_timeout_ms = example::get(config, "MSG_TIMEOUT_MS", GMSEC_WAIT_FOREVER);
		int prog_timeout_s = example::get(config, "PROG_TIMEOUT_S", GMSEC_WAIT_FOREVER);

		std::string subject = "GMSEC.TEST.PUBLISH.NOREPLY";
		if (subjects[0] != "GMSEC.>")
		{
			subject = subjects[0];
		}

		GMSEC_INFO << "Creating a request message";
		Message reqMsg(subject.c_str(), Message::REQUEST);

		GMSEC_INFO << "Registering event callback";
		connection->registerEventCallback(Connection::REQUEST_TIMEOUT_EVENT, &rcb);

		GMSEC_INFO << "Sending the request";
		connection->request(reqMsg, msg_timeout_ms, &rcb);
	
		if (prog_timeout_s == GMSEC_WAIT_FOREVER)
		{
			GMSEC_INFO << "Press the enter key to exit";
			std::cin.get();
		}
		else
		{
			example::millisleep(prog_timeout_s * 1000);
			GMSEC_INFO << "Program Timeout Reached!";
		}

		GMSEC_INFO << "Stopping AutoDispatch";
		connection->stopAutoDispatch();
	}
	catch (Exception& e)
	{
		GMSEC_ERROR << e.what();
		success = false;
	}

	return success;
}


int main(int argc, char* argv[])
{
	Config config(argc, argv);

	example::addToConfigFromFile(config);

	if (example::isOptionInvalid(config, argc))
	{
		example::printUsage("gmsub_disp_rr");
		return -1;
	}

	return (gmsub_disp_rr(config).run() ? 0 : -1);
}
